elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,419 @@
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-677[EAP7-677]
+* https://issues.redhat.com/browse/WFCORE-4584[WFCORE-4584] - Include the new wildfly-elytron-encryption module.
+* https://issues.redhat.com/browse/ELY-2055[ELY-2055] - Refactor CredentialStoreCommand to make development easier.
+* https://issues.redhat.com/browse/ELY-1850[ELY-1850] - CredentialStore API missing ability to check alias types.
+* https://issues.redhat.com/browse/ELY-2081[ELY-2081] - Develop a simple properties based credential store for secret keys.
+
+=== Other Links
+
+The following is a link to an early discussion on the mailto:wildfly-dev@lists.jboss.org[wildfly-dev] mailing list.
+
+https://lists.jboss.org/pipermail/wildfly-dev/2019-July/006932.html[Expression Resolver Discussion]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:okotek@redhat.com[Ondrej Kotek]
+
+=== Testing By
+
+[ ] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+Within the WildFly Elytron project new utilities will be added to handle the generation of `SecretKey` instances and the encryption and decryption of clear
+text values.  These utilities will support transfer of the values using versioned Base64 encoded tokens which will be described in more detail below.
+
+Within WildFly Core the general `ExpressionResolver` support will be updated to detect the availability of an expression resolver capability exposing
+a runtime API for expression resolution.
+
+Also within WildFly Core the Elytron subsystem will be updated with new resources as required, resources will also contain operations both for key management
+and operations for the encryption of clear text strings for use as expressions in the management model.
+
+Additionally all documentation lives within the WildFly project so updates will be applied there accordingly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The overall intent of this enhancement is for the Elytron subsystem to provide a configurable expression resolver resource to make use of an `AES`
+`SecretKey` loaded from a `CredentialStore` to decrypt expressions in the management model which have been previously encrypted using the same key.  To
+meet this requirement some additional functionality will also be required to hande the management of the keys.
+
+It is a requirement within this enhancement that multiple secret keys can be used for the decryption of expressions, these keys could be loaded from the same
+credential store or from separate credential stores - this relates to the "chicken and egg" problem which will be described later.
+
+The initial integration of this enhancement will not support specifying custom `java.security.Provider` instances as the key type and transformation will be
+using pre-defined defaults - however the expression resolver will be initialised after the Elytron subsystem has performed initialisation and registration of
```

Comment:
What about attributes of the Elytron subsystem and provider-loader that allow expressions? Do I understand it well that those will be processed just with the general resolver but the new credential store related resolver will not be applied to those?

Reply:
On this point there is going to be a limit to how far we can go.

Within the implementation the act of resolving an expression could trigger some initialisation - i.e. we need a credential store and we need the providers registered.  This initialisation could trigger the need for expression resolution.

We must be able to support the scenario where the clear-text password of one credential store is encrypted using the secret key of another credential store so if there is any recursion triggered for expression resolution that must be supported.

If however the recursion leads to an unresolvable cycle that will have to fail and lead to an error.  A cycle should be easy to detect i.e. if on first use of "resolverOne" the initialisation triggers something else requiring "resolverOne" we know we have a cycle - on a ThreadLocal we will likely maintain a Set of requested resolvers which would enable us to detect a longer chain forming a cycle.

If recursion is triggered alternative expression resolution using system properties should remain possible.

Reply:
Thanks a lot for the explanation. Maybe it's worth mentioning the solution here. In case you consider this just an implementation detail, this conversation can be resolved.

Reply:
I think this specific detail is an implementation detail, but this discussion has reminded me that we should list the relative-to paths that we can support.  As expression resolution is early we can not use user defined paths but we should support the standard set e.g. jboss.server.config.dir so I should list those.